components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/jwt/JWTValidator.java

```diff
@@ -0,0 +1,509 @@
+        String signatureAlgorithm;
+        JSONObject header;
+        try {
+            header = new JSONObject(new String(Base64.getUrlDecoder().decode(splitToken[0])));
+        } catch (JSONException | IllegalArgumentException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Token decryption failure when retrieving header. Token: " +
+                        getMaskedToken(splitToken), e);
+            }
+            throw new APISecurityException(APISecurityConstants.API_AUTH_INVALID_CREDENTIALS,
+                    "Invalid JWT token", e);
+        }
+        signatureAlgorithm = header.getString(APIConstants.JwtTokenConstants.SIGNATURE_ALGORITHM);
+        if (StringUtils.isBlank(signatureAlgorithm)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Signature algorithm not found in the token. Token: " + getMaskedToken(splitToken));
+            }
+            throw new APISecurityException(APISecurityConstants.API_AUTH_INVALID_CREDENTIALS, "Invalid JWT token");
+        }
+        if (APIConstants.SIGNATURE_ALGORITHM_RS256.equals(signatureAlgorithm)) {
+            signatureAlgorithm = APIConstants.SIGNATURE_ALGORITHM_SHA256_WITH_RSA;
+        }
+        return signatureAlgorithm;
+    }
+
+    /**
+     * Return tenant domain of the API being invoked.
+     *
+     * @return tenant domain
+     */
+    private String getTenantDomain() {
+        return PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+    }
+
+    private boolean isGatewayTokenCacheEnabled() {
+        try {
+            APIManagerConfiguration config = ServiceReferenceHolder.getInstance().getAPIManagerConfiguration();
+            String cacheEnabled = config.getFirstProperty(APIConstants.GATEWAY_TOKEN_CACHE_ENABLED);
+            return Boolean.parseBoolean(cacheEnabled);
+        } catch (Exception e) {
+            log.error("Did not found valid API Validation Information cache configuration. " +
+                    "Use default configuration.", e);
+        }
+        return true;
+    }
+
+    private Cache getGatewayTokenCache() {
+        return getCacheFromCacheManager(APIConstants.GATEWAY_TOKEN_CACHE_NAME);
+    }
+
+    private Cache getInvalidTokenCache() {
+        return getCacheFromCacheManager(APIConstants.GATEWAY_INVALID_TOKEN_CACHE_NAME);
+    }
+
+    private Cache getGatewayKeyCache() {
+        return getCacheFromCacheManager(APIConstants.GATEWAY_KEY_CACHE_NAME);
+    }
+
+    private Cache getCacheFromCacheManager(String cacheName) {
+        return Caching.getCacheManager(
+                APIConstants.API_MANAGER_CACHE_MANAGER).getCache(cacheName);
+    }
+
+    private String getAccessTokenCacheKey(String accessToken, String apiContext, String apiVersion,
+                                          String resourceUri, String httpVerb) {
+        return accessToken + ":" + apiContext + ":" + apiVersion + ":" + resourceUri + ":" + httpVerb;
+    }
```

Comment:
Constant can be used for ":"

Reply:
Fixed in https://github.com/wso2/carbon-apimgt/pull/9250